components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt

```diff
@@ -71,19 +71,22 @@ internal class EngineObserver(
         url: String,
         triggeredByRedirect: Boolean,
         triggeredByWebContent: Boolean,
-        shouldLoadUri: (Boolean) -> Unit
+        shouldLoadUri: (Boolean, String) -> Unit
     ) {
         if (triggeredByRedirect || triggeredByWebContent) {
             session.searchTerms = ""
         }
 
+        /* Debugging code for Android-components/issues/5127, will remove */
+        val shouldLoadUriWithId = { shouldLoad: Boolean -> shouldLoadUri(shouldLoad, this.toString()) }
```

Comment:
To really get to the bottom of the cause we also need to identify the session i.e.
- We could have multiple `EngineObserver`s on the same browser session (and therefore engine session) 
- We somehow have the same GeckoResult (or lambda) on multiple browser sessions

If we add the session ID to the string we'll know for sure what's happening. I am still unclear how any of this can actually happen though, but it doesn't hurt to get more clarity with this and help us find the cause.

Reply:
Should I'll add the Session ID in the caller info.   Thanks,